### PR TITLE
[FLINK-23960][streaming] checkScheduledTimestamps changed according to semantic of the calculation processing time delay 

### PR DESCRIPTION
**Pull request by akalash**
## What is the purpose of the change

*PR fixes the checkScheduledTimestamps test*

## Brief change log

  - *Removed usage of currentTimeMillis without a shift in checkScheduledTimestamps because it leaded to instability*

## Verifying this change

It is test itself

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ProcessingTimeServiceUtil.java [hunk 1]
```diff
@@ -26,21 +26,22 @@
...
-        //   timer by 1 ms to align the semantics with watermark. A watermark T says we won't see
-        //   elements in the future with a timestamp smaller or equal to T. With processing time, we
-        //   therefore need to delay firing the timer by one ms.
-        // (2) past timers(processingTimestamp < currentTimestamp): do not need to delay the firing
-        //   because currentTimestamp is larger than processingTimestamp pluses the 1ms offset.
+        // future/now timers(processingTimestamp >= currentTimestamp): delay the firing of the
+        // timer by 1 ms to align the semantics with watermark. A watermark T says we won't see
+        // elements in the future with a timestamp smaller or equal to T. Without this 1ms delay,
+        // if we had fired the timer for T at the timestamp T, it would be possible that we would
+        // process another record for timestamp == T in the same millisecond.
```
Comment (akalash) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ProcessingTimeServiceUtil.java:44]: I actually am not sure about these changes by myself so if somebody think that it isn't better than it was before we can just revert these changes.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ProcessingTimeServiceImpl.java [hunk 2]
```diff
@@ -61,7 +61,7 @@ public long getCurrentProcessingTime() {
     public ScheduledFuture<?> registerTimer(long timestamp, ProcessingTimeCallback target) {
         if (isQuiesced()) {
             return new NeverCompleteFuture(
-                    ProcessingTimeServiceUtil.getProcessingTimeDelay(
+                    ProcessingTimeServiceUtil.getRecordProcessingTimeDelay(
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ProcessingTimeServiceImpl.java:64]: I don't think this rename is a good idea. As I mentioned in the JIRA ticket, this doesn't have to be about processing records, but about any abstract events that are happening in the system using processing time.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ProcessingTimeServiceUtil.java [hunk 3]
```diff
@@ -26,21 +26,22 @@
 
     /**
      * Returns the remaining delay of the processing time specified by {@code processingTimestamp}.
+     * This delay should guarantee that the timer will be fired after all record with timestamp <=
+     * {@code processingTimestamp} have been already processed.
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ProcessingTimeServiceUtil.java:30]:
```
This delay guarantees that the timer will be fired at least 1ms after the time it's registered for.
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ProcessingTimeServiceUtil.java (hunk 1 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ProcessingTimeServiceUtil.java:44]: > it would be possible that we would process another record for timestamp == T in the same millisecond.

->

> it would be possible that we would process another record for timestamp == T in the same millisecond, but after the timer for the timsetamp T has already been fired.

Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/ProcessingTimeServiceUtil.java:44]: Plus I would keep the distinction between those two cases (the previous (1) and (2)).

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/StreamTaskTimerTest.java [hunk 4]
```diff
@@ -107,43 +109,42 @@ public void testErrorReporting() throws Exception {
...
-        while (errorRef.get() == null
-                && ValidatingProcessingTimeCallback.numInSequence < 4
-                && System.currentTimeMillis() < deadline) {
-            Thread.sleep(100);
+        ValidatingProcessingTimeCallback.numInSequence = 0;
+        long currentTimeMillis = System.currentTimeMillis();
+        ArrayList<ValidatingProcessingTimeCallback> timeCallbacks = new ArrayList<>();
+
+        /*
+         It is not possible to register timer for currentTimeMillis or value slightly greater than
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/StreamTaskTimerTest.java:117]: I believe that instead of:
> It is not possible to register timer for currentTimeMillis

You had this in mind?

> It is not possible to test registering timer for currentTimeMillis

?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/StreamTaskTimerTest.java [hunk 5]
```diff
@@ -107,43 +109,42 @@ public void testErrorReporting() throws Exception {
...
+        }
         assertEquals(4, ValidatingProcessingTimeCallback.numInSequence);
     }
 
     private static class ValidatingProcessingTimeCallback implements ProcessingTimeCallback {
 
         static int numInSequence;
 
-        private final AtomicReference<Throwable> errorRef;
+        private final CompletableFuture<Void> finished = new CompletableFuture<>();
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/StreamTaskTimerTest.java:142]: The bug fix in the test and the refactoring of this test I think are independent and should be in separate commits?

**Review by pnowojski — COMMENTED** [review]
Looks mostly Good To Me % a couple of minor comments.

**Review by pnowojski — APPROVED** [review]
Thanks for the updates!
